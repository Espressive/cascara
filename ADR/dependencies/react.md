# React

Our React version is tied to FE's version, we can bump only after they are doing bumping Ract. For now, all the following ecosystem members will be pinned to their current version:

- @types/react v16.8.0
- @types/react-dom v16.8.0
- react v16.13.1
- react-dom v16.13.1
- react-router v5.2.0
- react-router-dom v5.2.0
- react-scripts v4.0.3
- react-test-renderer v16.13.1
