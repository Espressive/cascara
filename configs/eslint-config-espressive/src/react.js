// "off" or 0 - turn the rule off
// "warn" or 1 - turn the rule on as a warning (doesn't affect exit code)
// "error" or 2 - turn the rule on as an error (exit code will be 1)

module.exports = {
  rules: {
    'react/boolean-prop-naming': 2,
    'react/default-props-match-prop-types': 2,
    'react/destructuring-assignment': 2,
    'react/jsx-fragments': 2,
    'react/jsx-no-duplicate-props': 2,
    'react/jsx-no-useless-fragment': 2,
    'react/jsx-pascal-case': 2,
    'react/jsx-sort-default-props': 2,
    'react/jsx-sort-props': 2,
    'react/no-children-prop': 2,
    'react/no-danger': 2,
    'react/no-find-dom-node': 2,
    'react/no-multi-comp': 2,
    'react/no-string-refs': 2,
    'react/no-this-in-sfc': 2,
    'react/no-typos': 2,
    'react/no-unused-prop-types': 2,
    'react/no-unused-state': 2,
    'react/prefer-stateless-function': 2,
    'react/self-closing-comp': 2,
    'react/sort-prop-types': 2,
    'react/style-prop-object': 2,
  },
};
