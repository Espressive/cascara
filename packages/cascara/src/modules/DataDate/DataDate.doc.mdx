---
title: Data Date
propTable: ./DataDate.js
type: 'module'
public: false
---

A data module to display **date** values, it renders a input of type **date**.

### Sandbox

Feel free to play around with the following code:

```jsx
<ModuleSandbox
  isEditing={true}
  record={{
    id: 1,
    firstName: 'John',
    lastName: 'Doe',
    date: '2021-07-22',
  }}
>
  <DataDate
    attribute='date'
    isEditable={true}
    label='Date'
    value='2021-07-22'
  />
</ModuleSandbox>
```
