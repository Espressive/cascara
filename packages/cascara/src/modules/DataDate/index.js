export { default, propTypes } from './DataDate';
