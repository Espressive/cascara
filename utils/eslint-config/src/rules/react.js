const { fixtures } = require('../utils/filePaths');

// For brevity, we start with the recomended defaults and extend with our own rules.
// Please make sure we are looking at any rules in the recommended set before adding
// them below. If we are overriding a recommended rule, we should make a comment for
// each rule.
//
// https://github.com/yannickcr/eslint-plugin-react

module.exports = {
  extends: ['plugin:react/recommended'],
  overrides: [
    {
      files: fixtures,
      rules: {
        'react/no-multi-comp': 'off',
        'react/prop-types': 'off',
      },
    },
  ],
  rules: {
    'react/boolean-prop-naming': 'error',
    'react/button-has-type': 'error',
    'react/default-props-match-prop-types': 'error',
    'react/destructuring-assignment': 'error',
    'react/forbid-foreign-prop-types': 'error',
    'react/forbid-prop-types': 'error',
    'react/function-component-definition': [
      'error',
      {
        namedComponents: 'arrow-function',
        unnamedComponents: 'arrow-function',
      },
    ],
    'react/jsx-boolean-value': 'error',
    'react/jsx-fragments': 'error',
    'react/jsx-key': 'error',
    'react/jsx-no-bind': 'error',
    'react/jsx-no-duplicate-props': 'error',
    'react/jsx-no-target-blank': 'error',
    'react/jsx-no-useless-fragment': 'error',
    'react/jsx-pascal-case': 'error',
    'react/jsx-sort-default-props': 'error',
    'react/jsx-sort-props': 'error',
    // 'react/jsx-wrap-multilines': ['error', {
    //   "arrow": "parens-new-line",
    //   "assignment": "parens-new-line",
    //   "condition": "ignore",
    //   "declaration": "parens-new-line",
    //   "logical": "ignore",
    //   "prop": "ignore",
    //   "return": "parens-new-line",
    // }],
    'react/no-access-state-in-setstate': 'error',
    'react/no-children-prop': 'error',
    'react/no-danger': 'error',
    'react/no-find-dom-node': 'error',
    'react/no-multi-comp': 'error',
    'react/no-render-return-value': 'error',
    'react/no-string-refs': 'error',
    'react/no-this-in-sfc': 'error',
    'react/no-typos': 'error',
    'react/no-unused-prop-types': 'error',
    'react/no-unused-state': 'error',
    'react/prefer-stateless-function': 'error',
    'react/require-render-return': 'error',
    'react/self-closing-comp': 'error',
    'react/sort-prop-types': 'error',
    'react/style-prop-object': 'error',
  },
};
