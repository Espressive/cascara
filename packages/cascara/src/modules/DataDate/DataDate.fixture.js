import React from 'react';
import DataDate from './DataDate';
import ModuleSandbox from '../ModuleSandbox';

const DisplayDataDate = ({ isEditing, ...rest }) => (
  <ModuleSandbox isEditing={isEditing}>
    <DataDate {...rest} />
  </ModuleSandbox>
);

const value = '2021-07-24';

const displayProps = {
  label: 'Display',
  value: value,
};

const editingProps = {
  isEditing: true,
  label: 'Editing',
  value,
};

// These can be used in tests
export { displayProps, editingProps };

// These are our fixtures
export default {
  display: <DisplayDataDate {...displayProps} />,
  displayNoLabel: <DisplayDataDate {...displayProps} isLabeled={false} />,
  editing: <DisplayDataDate {...editingProps} />,
  editingNoLabel: <DisplayDataDate {...editingProps} isLabeled={false} />,
};
