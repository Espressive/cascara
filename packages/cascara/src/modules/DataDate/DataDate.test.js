import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import DateDate from './DataDate';
import { DisplayDataDate } from './DataDate.fixture';

import fixtures from '../helpers/ModuleTestHelper.fixture';

const { checkbox: ModuleTestWrapper } = fixtures;

const VALUE = '2021-07-24';
const EXPECTED_VALUE_TYPE = 'string';
const DATA_DISPLAY = [
  {
    attribute: 'date',
    label: 'Date',
    module: 'date',
  },
];
const DATA = { date: VALUE };

describe('Input Date', () => {
  // without ModuleSandbox will render the property information into a span
  describe('default', () => {
    const testId = 'default';

    // We need a place to store the view for snapshot testing. This is not required when we are using `screen` directly from RTL.
    let view;

    beforeEach(() => {
      // Set the render container to our `view` so it is in scope for the snapshot test
      view = render(<DateDate data-testid={testId} />).container;
    });

    test('renders without any props', () => {
      expect(view).toMatchSnapshot();
    });

    test('renders a <span> by default', () => {
      const input = screen.getByTestId(testId);
      // Make sure the actual DOM element is not render an input
      expect(input.tagName).toMatch('SPAN');
    });
  });

  describe('editing', () => {
    // specifying ModuleSandbox will have the flag to render our input date field
    const testId = 'default';
    beforeEach(() =>
      render(
        <DisplayDataDate data-testid={testId} isEditing value={'2021-07-24'} />
      )
    );
    test('renders a <input date> by default', () => {
      const input = screen.getByTestId(testId);
      // Check that we also use the correct type for accessibility
      expect(input).toHaveAttribute('type', 'date');
      expect(input).toHaveValue(VALUE);
    });

    test('change value', () => {
      const newDate = '2021-07-23';
      const input = screen.getByTestId(testId);
      userEvent.type(input, newDate);
      expect(input).toHaveValue(newDate);
    });
  });

  // describe('data type', () => {
  //   test('it returns the correct data type', async () => {
  //     const handleFormAction = jest
  //       .fn()
  //       .mockImplementation((action, formData) => true);

  //     render(
  //       <ModuleTestWrapper
  //         data={DATA}
  //         dataDisplay={DATA_DISPLAY}
  //         isInitialEditing
  //         onAction={handleFormAction}
  //       />
  //     );

  //     await waitFor(() => screen.findByText('Date'));

  //     const dateInput = screen.getByText('Date');
  //     userEvent.click(dateInput);

  //     const saveButton = screen.getByRole('button', { name: 'Save' });
  //     userEvent.click(saveButton);

  //     expect(handleFormAction.mock.calls[0][0]).toEqual({ name: 'edit.save' });
  //     expect(handleFormAction.mock.calls[0][1]).toEqual({ date: VALUE });
  //     expect(typeof handleFormAction.mock.calls[0][1].date).toEqual(
  //       EXPECTED_VALUE_TYPE
  //     );
  //   });
  // });
});
