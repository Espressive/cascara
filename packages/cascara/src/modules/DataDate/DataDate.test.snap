// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataDate display snapshot 1`] = `
<div>
  <div
    class="Date"
  >
    <span>
      <span
        class="LabelText"
      >
        Display
      </span>
      <span
        aria-label="Display"
        class="Date"
      >
        2021-07-24
      </span>
    </span>
  </div>
</div>
`;

exports[`DataDate editing snapshot 1`] = `
<div>
  <div
    class="Date"
  >
    <label
      for="Editing"
    >
      <span
        class="LabelText"
      >
        Editing
      </span>
      <input
        class="Input"
        id="Editing"
        name="Editing"
        type="date"
        value="2021-07-24"
      />
    </label>
  </div>
</div>
`;
